### PR TITLE
Speaker dialog module shows sessions addressed by the speaker

### PR DESCRIPTION
**Pull request by SaptakS**
The speaker dialog box shows the sessions addressed by the speaker in card layout. Clicking on this sessions opens the dialog module for the particular session

### CODE REVIEW

File: app/components/speakers/SpeakersController.js [hunk 1]
```diff
@@ -77,6 +87,48 @@ speakersModule.controller('SpeakerDialogController',
                 {field: 'Position', value: sdc.speaker.position});
         }
 
+        sdc.count = function(speaker, sessions){
+            var count = 0;
+            for(var i = 0; i < sessions.length; i++) {
+                if(sessions[i].speakers.length > 0) {
```
Comment (creativepsyco) [inline app/components/speakers/SpeakersController.js]: should probably guard against speakers being `null`, just to be safe.

Comment (SaptakS) [inline app/components/speakers/SpeakersController.js]: OK. I didn't consider that. Will put that check.

File: app/components/speakers/SpeakersController.js [hunk 2]
```diff
@@ -77,6 +87,48 @@ speakersModule.controller('SpeakerDialogController',
...
+                    for(var j = 0; j < sessions[i].speakers.length; j++) {
+                        if(speaker.id == sessions[i].speakers[j].id)
+                            count++;
```
Comment (creativepsyco) [inline app/components/speakers/SpeakersController.js]: This function can be replaced by [Array:filter][1] 

```js
sessions.filter(function(session){
     return session.speakers.filter(function(sp){
              return sp.id == 291;
     })
    .length > 0;
})
```
e.g. to search for all sessions by speaker id 291. More concise and clear.

[1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter

File: app/components/speakers/SpeakersController.js [hunk 3]
```diff
@@ -77,6 +87,48 @@ speakersModule.controller('SpeakerDialogController',
...
+        
+        sdc.sessionsDetail = function(speaker, sessions) {
+            var k = 0;
+            var count = sdc.count(speaker,sessions);
+            var spSessions = new Array(count);
+            for(var i = 0; i < sessions.length;i++) {
+                if(sessions[i].speakers.length > 0) {
+                    for(var j = 0; j < sessions[i].speakers.length; j++) {
+                        if(speaker.id == sessions[i].speakers[j].id)
+                            spSessions[k++] = sessions[i];
```
Comment (creativepsyco) [inline app/components/speakers/SpeakersController.js]: can use the filter method here.

**Comment by SaptakS** [comment]
Sir, I have amended the code to use array filter. Please review.

File: app/components/speakers/SpeakersController.js [hunk 4]
```diff
@@ -77,6 +87,43 @@ speakersModule.controller('SpeakerDialogController',
...
         }
 
+        sdc.count = function(speaker, sessions){
+            var count = 0;
+            sessions.filter(function(session){
+                if (session.speakers.filter(function(sp){
+                          return sp.id == speaker.id;
+                 })
+                 .length > 0)
+                    count++;
```
Comment (creativepsyco) [inline app/components/speakers/SpeakersController.js]: The filter function should not be used to do calculations inside. I suppose you are interested in counting the number of sessions by the particular speaker. Then you can just use:

```js
var matchingSessions = sessions.filter(function(session){return session.speakers.filter(function(sp){return sp.id == speaker.id}).length > 0;})

size = matchingSessions.length
```

You can always assume that you will get array back, the filter function must return `true` or `false` and should not do any calculation inside.

File: app/components/speakers/SpeakersController.js [hunk 5]
```diff
@@ -77,6 +87,43 @@ speakersModule.controller('SpeakerDialogController',
                 {field: 'Position', value: sdc.speaker.position});
         }
 
+        sdc.count = function(speaker, sessions){
+            var count = 0;
+            sessions.filter(function(session){
```
Comment (creativepsyco) [inline app/components/speakers/SpeakersController.js]: missing return statement here.

**Comment by creativepsyco** [comment]
@SaptakS see my comments above.

File: app/components/speakers/SpeakersController.js (hunk 4 above)
Comment (SaptakS) [inline app/components/speakers/SpeakersController.js]: Ok sir.

**Comment by SaptakS** [comment]
Amended the changes. Please review.

**Comment by creativepsyco** [comment]
:+1: Looks good thanks for the fix.

**Comment by SaptakS** [comment]
(y)

**Pull request merged**